src/toolbars/TimeToolBar.cpp

```diff
@@ -60,6 +61,10 @@ TimeToolBar::TimeToolBar(AudacityProject &project)
 {
    mSubscription =
       ProjectRate::Get(project).Subscribe(*this, &TimeToolBar::OnRateChanged);
+
+   mTimeSignatureChangedSubscription =
+      ProjectTimeSignature::Get(project).Subscribe([this](auto)
+                                                   { UpdateTimeSignature(); });
```

Comment:
At last I have reached the commit in which `NumericConverterConfig`, introduced five commits previously, is fully used.

Forget what I said about `std::any`, yet I'm still uncomfortable with that structure and they messages are passed.

Mostly it's analogous with how changes of project rate are communiated to the formatter -- and I have decided I dislike that too.

Figuring out the more pleasing organization looks like it will turn into another proposed commit from me insted of an explanation, because I wanted to make the change myself to see if it made sense and compile, and it became rather extensive.

Reply:
I've planned to change that to AudacityProject reference, if you don't mind.

Reply:
And I plan to move beat/bar calculation to the ProjectTimeSignature

Reply:
> I've planned to change that to AudacityProject reference, if you don't mind.

I approve.

So no more `NumericConveterConfig`?

AudacityProject is at a low level of the dependency tree and you can think of it as the context argument for many things that need per-project state, and of course attaching more stuff non-intrusively to the project is what `ClientData` is all about.

All that was very much not true of the source code tree before 2019.  The Project depended on almost everything else becuase it had accessor functions defined in it to retrieve all those others.  This was a major reason why hundreds of source files were in a big strongly connected component of the graph.

Inventing `ClientData` was a great exercise of the Dependency Inversion Principle.  I didn't know the SOLID jargon at the time actually but I was led to the principle by experience.